### PR TITLE
better diagnostics for use_shared_state

### PR DESCRIPTION
**Pull request by Niedzwiedzw**
Currently it is quite hard to debug what causes the panics at runtime, this is an attempt making it a bit less intimidating for new users.
I also propose exposing the `try_*` methods in case someone wanted to handle the errors somehow (current implementation crashes and leaves the application frozen which might be really confusing to end user)

changelist: 
 - expose `try_*` methods on `UseSharedState<T>`
 - add meaningful panic information on `read()`/`write()` errors
 - unify `thiserror`'s versions across all packages and make it a `workspace`-wide dependency
 - above mentioned `read()` / `write()` methods also track callers in debug mode, providing specific location at which borrow error occurred in the panic message

### CODE REVIEW

File: packages/hooks/src/use_shared_state.rs
Comment (Demonthos) [inline packages/hooks/src/use_shared_state.rs:125]: It would be nice to provide information about the location where the state was previously read and written to in debug mode to provide nicer debug information. (This would use https://doc.rust-lang.org/reference/attributes/codegen.html#the-track_caller-attribute to track the file location of the caller of read and write in debug mode)

(It doesn't need to implemented in this PR)

Comment (Niedzwiedzw) [reply packages/hooks/src/use_shared_state.rs:126]: oooh I've never seen that one - nice. (pushed)

Comment (Niedzwiedzw) [reply packages/hooks/src/use_shared_state.rs]: pushed another one, so that it's only there in `debug` mode

Comment (Niedzwiedzw) [reply packages/hooks/src/use_shared_state.rs]: oh yeah, tracking itself should also be debug-only - pushed

Comment (Demonthos) [reply packages/hooks/src/use_shared_state.rs]: Apparently what I was thinking of is already implemented in https://docs.rs/debug-cell/latest/debug_cell. It tracks the current borrows and logs the location of any relevant borrows on errors

Comment (Niedzwiedzw) [reply packages/hooks/src/use_shared_state.rs]: hmmm this could almost be a separate feature behind a specific feature flag, idk your call

Comment (Demonthos) [reply packages/hooks/src/use_shared_state.rs]: It is probably better implemented in a different PR, but I think we could enable it in debug mode by default

Comment (Niedzwiedzw) [reply packages/hooks/src/use_shared_state.rs]: want me to do it?

Comment (Demonthos) [reply packages/hooks/src/use_shared_state.rs]: If you are up for it, that would be great! We could also use the same method for the use_ref hook

Comment (Niedzwiedzw) [reply packages/hooks/src/use_shared_state.rs]: ok soo I've done something a bit different instead but with the same result... let me know what you think

Comment (Niedzwiedzw) [reply packages/hooks/src/use_shared_state.rs]: soo the problem is... it doesn't work

```rust
Writing to shared state failed: [/home/niedzwiedz/.cargo/git/checkouts/dioxus-7f2b82817ca79561/760d937/packages/hooks/src/use_shared_state.rs:296:27] dioxus_hooks::use_shared_state::UseSharedState<editor::app::App> is already borrowed at [None], so it cannot be borrowed mutably.
```

as you can see it shows the spot inside `hooks/src/use_shared_state.rs` ...

Comment (Niedzwiedzw) [reply packages/hooks/src/use_shared_state.rs]: ok it doesn't actually work, it only shows the thing that was previously borrowed but it could've been dropped in the meantime. I guess there's no way around it other than rolling a custom RefCell that tracks borrows but also when they are released

Comment (Niedzwiedzw) [reply packages/hooks/src/use_shared_state.rs]: https://github.com/alexcrichton/debug-cell/pull/5 made a PR to `debug-cell` to have it expose the `try_*` methods on debug RefCell, let's hope it's not too much to ask 🤞🏿

**Review by Demonthos — APPROVED** [review]
This looks great! The new error is so much more helpful.

We need to either wait for the PR you opened to be merged or publish a forked version before this PR is merged. (I don't think we can have git dependancies in the crates.io version of Dioxus)

**Comment by Niedzwiedzw** [comment]
I would suggest forking debug-cell into DioxusLab for the time being, this repository doesn't have much traffic sadly

**Comment by Niedzwiedzw** [comment]
@jkelleyrtp  it still points to my fork of debug-cell... is this a good idea? I don't want to set a precedence like that...

**Pull request merged**
